schema/chat_messages.go

```diff
@@ -84,26 +117,44 @@ type ChatResult struct {
 
 // GetBufferString gets the buffer string of messages.
 func GetBufferString(messages []ChatMessage, humanPrefix string, aiPrefix string) (string, error) {
-	stringMessages := []string{}
+	result := []string{}
 	for _, m := range messages {
-		var role string
-		switch m.GetType() {
-		case ChatMessageTypeHuman:
-			role = humanPrefix
-		case ChatMessageTypeAI:
-			role = aiPrefix
-		case ChatMessageTypeSystem:
-			role = "System"
-		case ChatMessageTypeGeneric:
-			cgm, ok := m.(GenericChatMessage)
-			if !ok {
-				return "", fmt.Errorf("%w -%+v", ErrUnexpectedChatMessageType, m)
+		role, err := getMessageRole(m, humanPrefix, aiPrefix)
+		if err != nil {
+			return "", err
+		}
+		msg := fmt.Sprintf("%s: %s", role, m.GetContent())
+		if m, ok := m.(AIChatMessage); ok && m.FunctionCall != nil {
+			j, err := json.Marshal(m.FunctionCall)
+			if err != nil {
+				return "", err
 			}
-			role = cgm.Role
-		default:
-			return "", ErrUnexpectedChatMessageType
+			msg = fmt.Sprintf("%s %s", msg, string(j))
+		}
+		result = append(result, msg)
+	}
+	return strings.Join(result, "\n"), nil
+}
+
+func getMessageRole(m ChatMessage, humanPrefix, aiPrefix string) (string, error) {
```

Comment:
Is there a specific reason to not make a method on `ChatMessage`?

Reply:
Chat message is an interface.